types/core.go

```diff
@@ -0,0 +1,335 @@
+	"github.com/kong/go-kong/kong"
+)
+
+func eventFromArg(arg crud.Arg) crud.Event {
+	event, ok := arg.(crud.Event)
+	if !ok {
+		panic("unexpected type, expected diff.Event")
+	}
+	return event
+}
+
+type Entity interface {
+	Type() string
+	CRUDActions() crud.Actions
+	PostProcessActions() crud.Actions
+}
+
+type entityImpl struct {
+	typ                string
+	cRUDActions        crud.Actions // needs to set client
+	postProcessActions crud.Actions // needs currentstate Set
+}
+
+func (e entityImpl) Type() string {
+	return e.typ
+}
+
+func (e entityImpl) CRUDActions() crud.Actions {
+	return e.cRUDActions
+}
+
+func (e entityImpl) PostProcessActions() crud.Actions {
+	return e.postProcessActions
+}
+
+type EntityOpts struct {
+	CurrentState  *state.KongState
+	TargetState   *state.KongState
+	KongClient    *kong.Client
+	KonnectClient *konnect.Client
+}
+
+const (
+	// Service identifies a Service in Kong.
+	Service = "service"
+	// Route identifies a Route in Kong.
+	Route = "route"
+	// Plugin identifies a Plugin in Kong.
+	Plugin = "plugin"
+
+	// Certificate identifies a Certificate in Kong.
+	Certificate = "certificate"
+	// SNI identifies a SNI in Kong.
+	SNI = "sni"
+	// CACertificate identifies a CACertificate in Kong.
+	CACertificate = "ca-certificate"
+
+	// Upstream identifies a Upstream in Kong.
+	Upstream = "upstream"
+	// Target identifies a Target in Kong.
+	Target = "target"
+
+	// Consumer identifies a Consumer in Kong.
+	Consumer = "consumer"
+	// ACLGroup identifies a ACLGroup in Kong.
+	ACLGroup = "acl-group"
+	// BasicAuth identifies a BasicAuth in Kong.
+	BasicAuth = "basic-auth"
+	// HMACAuth identifies a HMACAuth in Kong.
+	HMACAuth = "hmac-auth"
+	// JWTAuth identifies a JWTAuth in Kong.
+	JWTAuth = "jwt-auth"
+	// MTLSAuth identifies a MTLSAuth in Kong.
+	MTLSAuth = "mtls-auth"
+	// KeyAuth identifies aKeyAuth in Kong.
+	KeyAuth = "key-auth"
+	// OAuth2Cred identifies a OAuth2Cred in Kong.
+	OAuth2Cred = "oauth2-cred"
+
+	// RBACRole identifies a RBACRole in Kong Enterprise.
+	RBACRole = "rbac-role"
+	// RBACEndpointPermission identifies a RBACEndpointPermission in Kong Enterprise.
+	RBACEndpointPermission = "rbac-endpoint-permission"
+
+	// ServicePackage identifies a ServicePackage in Konnect.
+	ServicePackage = "service-package"
+	// ServiceVersion identifies a ServiceVersion in Konnect.
+	ServiceVersion = "service-version"
+	// Document identifies a Document in Konnect.
+	Document = "document"
+)
+
+// AllTypes represents all types defined in the
+// package.
+var AllTypes = []string{
+	Service, Route, Plugin,
+
+	Certificate, SNI, CACertificate,
+
+	Upstream, Target,
+
+	Consumer,
+	ACLGroup, BasicAuth, KeyAuth,
+	HMACAuth, JWTAuth, OAuth2Cred,
+	MTLSAuth,
+
+	RBACRole, RBACEndpointPermission,
+
+	ServicePackage, ServiceVersion, Document,
+}
+
+func NewEntity(t string, opts EntityOpts) (Entity, error) {
+	switch t {
+	case Service:
+		return entityImpl{
+			typ: "service",
+			cRUDActions: &serviceCRUD{
```

Comment:
Minor readability thing, but case should be consistent across the acronym, i.e. `crudActions` rather than `cRUDActions`. If the first letter needs to be lowercase to not expose it, the rest should be also. "CRUD" is a common-enough acronym that it's recognizable as such even when not capitalized, thankfully.

Reply:
+1 to that, there's an explicit recommendation in the canon:

https://github.com/golang/go/wiki/CodeReviewComments#initialisms